File: packages/java/src/test/java/com/designpatterns/PatternsTest.java

package com.designpatterns;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

class PatternsTest {
    @Test
    void exportsAllPatterns() {
        assertEquals(23, Patterns.ALL.size());
    }

    @Test
    void filtersByQueryAndCategory() {
        List<Pattern> filtered = PatternFilter.filterPatterns(Patterns.ALL, "global access", "Creational");
        assertEquals(1, filtered.size());
        assertEquals("singleton", filtered.get(0).getSlug());
    }

    @Test
    void matchesQueryCaseInsensitively() {
        assertTrue(PatternFilter.matchesQuery(Patterns.ALL.get(18), "PUB/SUB"));
    }

    @Test
    void countsByCategory() {
        Map<String, Integer> counts = PatternFilter.countByCategory(Patterns.ALL);
        assertEquals(5, counts.get("Creational"));
        assertEquals(7, counts.get("Structural"));
        assertEquals(11, counts.get("Behavioral"));
    }
}

File: packages/go/designpatterns/patterns_test.go
package designpatterns

import "testing"

func TestPatterns(t *testing.T) {
	if len(Patterns) != 23 {
		t.Fatalf("expected 23 patterns, got %d", len(Patterns))
	}
}

func TestFilterPatterns(t *testing.T) {
	filtered := FilterPatterns(Patterns, "global access", "Creational")
	if len(filtered) != 1 || filtered[0].Slug != "singleton" {
		t.Fatalf("unexpected filter result: %#v", filtered)
	}
}

func TestMatchesQuery(t *testing.T) {
	if !MatchesQuery(Patterns[18], "PUB/SUB") {
		t.Fatal("expected observer pattern to match pub/sub query")
	}
}

func TestCountByCategory(t *testing.T) {
	counts := CountByCategory(Patterns)
	if counts["Creational"] != 5 || counts["Structural"] != 7 || counts["Behavioral"] != 11 {
		t.Fatalf("unexpected counts: %#v", counts)
	}
}
